src/json.ts

```diff
@@ -100,10 +100,25 @@ export const JsonRpcErrorStruct = object({
   stack: optional(string()),
 });
 
+/**
+ * Mark a certain key of a type as optional.
+ */
+export type OptionalField<
+  Type extends Record<string, unknown>,
+  Key extends keyof Type
+> = Omit<Type, Key> & Partial<Pick<Type, Key>>;
+
 /**
  * A JSON-RPC error object.
+ *
+ * Note that TypeScript infers `unknown | undefined` as `unknown`, meaning that
+ * the `data` field is not optional. To make it optional, we use the
+ * `OptionalField` helper, to explicitly make it optional.
  */
-export type JsonRpcError = Infer<typeof JsonRpcErrorStruct>;
+export type JsonRpcError = OptionalField<
```

Comment:
This will still be a problem when using `JsonRpcResponse`, we should make sure that is not the case.